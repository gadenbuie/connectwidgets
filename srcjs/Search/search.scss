.rscsearch {
  align-items: center;
  display: flex;
  position: relative;

  &::before {
    background-image: url("data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTciIGhlaWdodD0iMTciIHZpZXdCb3g9IjAgMCAxNyAxNyIgZmlsbD0iIzQ0NDQ0NCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48cGF0aCBkPSJNMTAuOTc5MiA5LjkxNjY3SDEwLjQxOTZMMTAuMjIxMiA5LjcyNTQyQzEwLjkxNTQgOC45MTc5MiAxMS4zMzMzIDcuODY5NTggMTEuMzMzMyA2LjcyOTE3QzExLjMzMzMgNC4xODYyNSA5LjI3MjA4IDIuMTI1IDYuNzI5MTcgMi4xMjVDNC4xODYyNSAyLjEyNSAyLjEyNSA0LjE4NjI1IDIuMTI1IDYuNzI5MTdDMi4xMjUgOS4yNzIwOCA0LjE4NjI1IDExLjMzMzMgNi43MjkxNyAxMS4zMzMzQzcuODY5NTggMTEuMzMzMyA4LjkxNzkyIDEwLjkxNTQgOS43MjU0MiAxMC4yMjEyTDkuOTE2NjcgMTAuNDE5NlYxMC45NzkyTDEzLjQ1ODMgMTQuNTEzN0wxNC41MTM3IDEzLjQ1ODNMMTAuOTc5MiA5LjkxNjY3Wk02LjcyOTE3IDkuOTE2NjdDNC45NjU0MiA5LjkxNjY3IDMuNTQxNjcgOC40OTI5MiAzLjU0MTY3IDYuNzI5MTdDMy41NDE2NyA0Ljk2NTQyIDQuOTY1NDIgMy41NDE2NyA2LjcyOTE3IDMuNTQxNjdDOC40OTI5MiAzLjU0MTY3IDkuOTE2NjcgNC45NjU0MiA5LjkxNjY3IDYuNzI5MTdDOS45MTY2NyA4LjQ5MjkyIDguNDkyOTIgOS45MTY2NyA2LjcyOTE3IDkuOTE2NjdaIiBmaWxsPSIjNDQ0NDQ0Ii8+PC9zdmc+");
    background-position: center center;
    background-repeat: no-repeat;
    content: " ";
    height: 3em;
    position: absolute;
    left: 0;
    top: 0;
    width: 2.7em;
  }

  &__input {
    border-radius: 3px;
    box-sizing: border-box;
    height: 3em;
    padding: 0 0.7em 0 2.7em;
    max-width: 100%;
    
    &:focus {
      outline: none;
    }
  }
}