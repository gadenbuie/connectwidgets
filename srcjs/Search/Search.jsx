/**
 * rscsearch main component
 * React provided at window via reactR
 */

import { cloneDeep } from 'lodash';
import PropTypes from 'prop-types';
import { debounce } from '@/utils';
import './search.scss';

class Search extends React.Component {
  constructor(props) {
    super(props);
    const crosstalkHandle = new crosstalk.FilterHandle();
    crosstalkHandle.setGroup(props.crosstalkGroup);

    const dataKeys = Array.isArray(props.crosstalkKey)
                      ? props.crosstalkKey
                      : props.data[props.crosstalkKey];

    this.state = {
      data: cloneDeep(props.data),
      dataKeys,
      crosstalkHandle,
    };

    this.search = debounce(300, this.search.bind(this));
    this.onInputChange = this.onInputChange.bind(this);
  }

  onInputChange(ev) {
    const prefix = ev.target.value.trim();
    this.search(prefix);
  }

  search(prefix) {
    if (prefix === "") {
      this.state.crosstalkHandle.clear();
      return;
    }

    const regx = new RegExp(prefix, "i");
    const results = this.state.data.title.reduce((matching, title, index) => {
      if (regx.test(title) || regx.test(this.state.data.name[index])) {
        // dataframe rows start at 1 and aren't based on index 0
        matching.push(this.state.dataKeys[index]);
      }
      return matching;
    }, []);

    this.state.crosstalkHandle.set(results);
  }

  render() {
    return (
      <div className="rscsearch">
        <input
          className="rscsearch__input"
          type="text"
          placeholder="Search"
          onChange={this.onInputChange}
        />
      </div>
    );
  }
}

Search.propTypes = {
  data: PropTypes.object,
  crosstalkKey: PropTypes.array,
  crosstalkGroup: PropTypes.string,
}

export default Search;
