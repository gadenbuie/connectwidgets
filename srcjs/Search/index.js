import { reactWidget } from 'reactR';
import Search from './Search';

reactWidget('rsc_search', 'output', { Search });
