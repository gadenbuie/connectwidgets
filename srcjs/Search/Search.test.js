import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { range } from 'lodash';
import Search from './Search';

jest.useFakeTimers();

const testFrameData = {
  // Search component only cares about name and title cols.
  name: [
    'cars-rmd-123456',
    'woodfillers-rmarkdown-123456',
    'shiny-app-987654',
    'shiny-app-159870',
    'mood-rmd-561237',
    '',
  ],
  title: [
    'Cars Report 2021',
    'Woodfillers resistance report',
    '',
    'Drying tomatoes results',
    'Colors impact in mood',
    'Research by Tommy',
  ],
};

const fulfillDebounce = () => jest.runAllTimers();

const inputEvWith = value => ({ target: { value }});

const mkWrapper = () => {
  return shallow(
    <Search
      data={testFrameData}
      crosstalkKey={range(1, testFrameData.name.length + 1)}
      crosstalkGroup="abcd-1234"
    />
  );
};

describe('<Search />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders properly, setup crosstalk handle', () => {
    const wrapper = mkWrapper();
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(crosstalk.FilterHandle).toHaveBeenCalled();
    expect(crosstalk.FilterHandle().setGroup).toHaveBeenCalledWith('abcd-1234');
  });

  describe('while searching', () => {
    it('searching emits via crosstalk when typing debounce fullfills', () => {
      const wrapper = mkWrapper();
  
      // Won't be called immediately due to debounce
      wrapper.find('input').simulate('change', inputEvWith('horses'));
      expect(crosstalk.FilterHandle().set).not.toHaveBeenCalled();
      
      // NO rows match 'horses'
      fulfillDebounce();
      expect(crosstalk.FilterHandle().set).toHaveBeenCalledWith([]);
    });

    it.each([
      ['shiny', [3, 4]],
      ['Tom', [4, 6]],
      ['tommy', [6]],
      ['report', [1, 2]],
      ['vegetables', []],
    ])('"%s", expects rows %s', (term, expectedRows) => {
      const wrapper = mkWrapper();
      wrapper.find('input').simulate('change', inputEvWith(term));
      fulfillDebounce();
      expect(crosstalk.FilterHandle().set).toHaveBeenCalledWith(expectedRows);
    });

    it('crosstalk filter clears with empty search term', () => {
      const wrapper = mkWrapper();
  
      wrapper.find('input').simulate('change', inputEvWith('horses'));
      fulfillDebounce();
      expect(crosstalk.FilterHandle().set).toHaveBeenCalledWith([]);

      wrapper.find('input').simulate('change', inputEvWith(''));
      fulfillDebounce();
      expect(crosstalk.FilterHandle().clear).toHaveBeenCalled();
    });
  });
});
