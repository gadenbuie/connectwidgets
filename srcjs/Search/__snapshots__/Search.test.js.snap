// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> renders properly, setup crosstalk handle 1`] = `
<div
  className="rscsearch"
>
  <input
    className="rscsearch__input"
    onChange={[Function]}
    placeholder="Search"
    type="text"
  />
</div>
`;
